### PR TITLE
Closes #388 [2]

### LINKED ISSUE #388: Sandbox security escalation allowing users to read configuration
I recently discovered that when using the sandbox of Twig, the global variables added by Laravel are still available. By default, [the framework adds 'app' as a shared variable in the ViewServiceProvider](https://github.com/laravel/framework/blob/master/src/Illuminate/View/ViewServiceProvider.php#L52).

The problem with this is that although by itself app, as an instance of Application is quite harmless, the [Container instance which it inherits is not](https://github.com/laravel/framework/blob/master/src/Illuminate/Container/Container.php#L1279).
Twig's sandbox requires you to add the magic `__isset()` method to your classes in order to allow the sandbox' security policy to check if the method or property you are calling on your class is allowed. Laravel's Container class does not have this method, meaning that the `__get()` method can be used freely within the sandbox, regardless of the given security policy.

The Application class contains all instances for the application, including the configuration one. This configuration can be requested by simply calling `$application->config`, which will use the magic get method to fetch the config repository, which in turn allows you to request any value from the current configuration [as it implements ArrayAccess](https://github.com/twigphp/Twig/issues/2878) (unfortunately #362 doesn't fix this as it's not a Model).

## Tl;dr
You can read the configuration of any Laravel application from within Twig's sandbox by using the global app variable added by the framework itself. 

For example:
```twig
{% sandbox %}
    {% include(template_from_string(user_template) only %}
{% endsandbox %}
```
with `user_template` being the following:
```twig
The key is {{ app.config.app.key }}
```
would produce `The key is base64:your_app_key...`, while using the `config_get()` function would result in a Twig SecurityException being thrown as intended.

I don't suppose many users of this package are even using the sandbox or care about this issue but my proposed solution for now is to remove the `app` variable from the global context by either overriding it locally for your template or from within one of your app's ServiceProviders:
```twig
{% sandbox %}
    {% include(template_from_string(user_template) with {app: null } only %}
{% endsandbox %}
```
Or in a ServiceProvider:
```php
public function boot(): void {
    View::share('app', null);
}
```

Again, I don't know how big of an issue this really is with this package itself but I thought I'd leave it here anyway so it can either be documented somewhere or discussed.

### PR DESCRIPTION
**Pull request by bytestream**
Remake of https://github.com/rcrowe/TwigBridge/pull/398 which was reverted due to https://github.com/rcrowe/TwigBridge/pull/398#issuecomment-653434596

Before we were use checking for calls against Eloquent models. #388 is lack of sandbox checks on Container objects. 

The problem with the old PR was that it was checking `ArrayAccess` but using `$object->$item`  - see https://github.com/rcrowe/TwigBridge/pull/398#issuecomment-653512286

This PR resolves that issue by using `$object[$item]` and also adds the `isset()` check which happens in the upstream code - https://github.com/twigphp/Twig/blob/3.x/src/Extension/CoreExtension.php#L1370

### CODE REVIEW

**Comment by bytestream** [comment]
This should resolve your issue with the old PR @barryvdh ?

**Comment by bytestream** [comment]
@barryvdh this version does the trick. You can compare https://github.com/rcrowe/TwigBridge/pull/401/files vs https://github.com/rcrowe/TwigBridge/pull/398/files

The key part is the `isset` check that I mentioned was probably missing from the old PR. That's what caused your issue.

**Comment by ndhan93** [comment]
I think this shouldn't enable by default, this extension just implement sandbox checking for ArrayAccess (also I think you should rename this extension)
If twig not support this feature, we shouldn't change it

For issue #388 we need remake app variable like this: https://github.com/symfony/symfony/blob/master/src/Symfony/Bridge/Twig/AppVariable.php

**Comment by barryvdh** [comment]
So in sandbox, just add that object instead?

**Comment by barryvdh** [comment]
Just to be clear, does this PR impact non-sandbox behavior?

**Comment by bytestream** [comment]
Can you justify why we shouldn't support it? Technically twig doesn't sorry Eloquent model checks, hence the override... it doesn't seem like a far stretch to extend this to all arrayaccess usage hidden how wide spread it is in Laravel

I would be ok for your to be being a feature flag in the config if that pleases people though...

**Comment by ndhan93** [comment]
> So in sandbox, just add that object instead?

we apply it global, not sandbox only
replace app variable with custom object, not using laravel application object
same as Symfony, app variable only wrap request, session, user (or any SAFE object, not ALL)
@barryvdh what do you think ?

> Can you justify why we shouldn't support it? Technically twig doesn't sorry Eloquent model checks, hence the override... it doesn't seem like a far stretch to extend this to all arrayaccess usage hidden how wide spread it is in Laravel
> 
> I would be ok for your to be being a feature flag in the config if that pleases people though...

in #362 he said call ``{{model.save}}`` will execute ``$model->save()``, that is bug in laravel 5.2 only, they fixed it, so we don't care that problem right now
ArrayAccess in laravel eloquent only access to attributes, getter and relationship (eloquent again), I don't see any risk with that
my opinion, this package is bridge, not customize twig for laravel, if we make any breaking change, it should be optional

**Comment by bytestream** [comment]
Can you provide a link to the PR that fixes that in Eloquent?

If it is fixed then I don't mind the approach you're suggesting. It removes a lot of unnecessary complexity from this package. The only problem... are there people on v0.12 of this package that are using a Laravel version where it's not fixed upstream?

**Comment by bytestream** [comment]
It looks like current branch supports Laravel 5.5 or above. So is this fixed in Laravel 5.5 and above? Link to PR would be good.

**Comment by ndhan93** [comment]
this commit: https://github.com/laravel/framework/commit/7ce97bec546d607ca207082722ceecabfbcc11a9#diff-d8e24a5815a26f7211e66db787be647f

**Comment by bytestream** [comment]
Cool ta. Looks like we can get rid of the node visitors and overriding twig_get_attribute functionality. 

Do you want to have a go at what you're suggesting @ndhan93 ?

**Comment by ndhan93** [comment]
> Cool ta. Looks like we can get rid of the node visitors and overriding twig_get_attribute functionality.
> 
> Do you want to have a go at what you're suggesting @ndhan93 ?

your extension so cool, I think many people will need it
I think you should rename your extension to make it sense (eg: ArrayAccessSandboxCheck), and comment it in config file
I'll make PR to fix issue #388 right now

**Comment by bytestream** [comment]
@barryvdh would you mind tagging this?

**Comment by jshah4517** [comment]
@barryvdh would it be possible to have a new release containing this please?

**Comment by ovp87** [comment]
Any resolution to this? all solutions appear to have been reverted.
